lib/serviceDiscovery/helpers.js

```diff
@@ -0,0 +1,52 @@
+'use strict';
+
+const os = require('os');
+const crypto = require('crypto');
+const memoizee = require('memoizee');
+
+const interfaces = os.networkInterfaces();
+const ifaceNames = Object.keys(interfaces);
+
+// choose the first valid ip as the one we will announce
+let announceIps;
+
+/**
+ * Builds a list of every announce-able ips on the server
+ *
+ * @returns {string} The ip list
+ * @todo Support from env variables?
+ */
+exports.getAnnounceIps = memoizee(function () {
+	if (announceIps) {
+		return announceIps;
+	}
+
+	announceIps = [];
+
+	for (var i = 0; i < ifaceNames.length; i++) {
+		var addresses = interfaces[ifaceNames[i]];
+
+		for (var j = 0; j < addresses.length; j++) {
+			var address = addresses[j];
+
+			// skip anything that doesn't interest us
+			if (address.internal) {
+				continue;
+			}
+
+			// found it, store it and break
+			announceIps.push(address.address);
+		}
+	}
+
+	// just in case
+	return announceIps;
+});
+
+exports.generateId = function (port) {
+	const generator = crypto.createHmac('sha256', 'secret');
+	exports.getAnnounceIps().forEach((ip) => generator.update(ip));
```

Comment:
I understand you're trying to get this ID to be consistent for the machine this is running on, but since we're doing `Object.keys` on line 8 and not sorting, I don't think order can be guaranteed. It would probably be wise to add a `.sort()` on this line:

```js
exports.getAnnounceIps().slice().sort().forEach((ip) => generator.update(ip));
```

(turns that code a bit ugly now, but ... I hope it illustrates the point)

Reply:
Good catch. I'll sort in getAnnounceIps instead though.

Reply:
Sorting there actually ends up having an adverse effect. See https://github.com/mage/mage/pulls for more details.